### PR TITLE
add NET7 to CI and tests

### PR DESCRIPTION
**Pull request by TimothyMothra**
Towards #732 

## Changes
- Any workflow that had ONLY NET6 was changed to NET7
- Any workflow that had NET6, I added NET7
- Any test csproj that had NET6, I added NET7
- Reorder test TargetFrameworks in descending order
- Add new docker-compose file for net7

## Todo
- will need to modify branch rules to require net7 builds

### CODE REVIEW

**Comment by TimothyMothra** [comment]
The test `AttachLogsToActivityEventTest` is failing. Need more time to investigate.

```
Message: 
System.Collections.Generic.KeyNotFoundException : The given key 'scope[0].NodeId' was not present in the dictionary.

  Stack Trace: 
Dictionary`2.get_Item(TKey key)
ActivityEventAttachingLogProcessorTests.AttachLogsToActivityEventTest(Boolean sampled, Int32 eventId, Boolean includeFormattedMessage, Boolean parseStateValues, Boolean includeScopes, Boolean recordException) line 151
RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

File: test/OpenTelemetry.Instrumentation.EventCounters.Tests/OpenTelemetry.Instrumentation.EventCounters.Tests.csproj [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp3.1;net6.0</TargetFrameworks>
+    <TargetFrameworks>netcoreapp3.1;net6.0;net7.0</TargetFrameworks>
```
Comment (cijothomas) [inline test/OpenTelemetry.Instrumentation.EventCounters.Tests/OpenTelemetry.Instrumentation.EventCounters.Tests.csproj:4]: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/OpenTelemetry.csproj#L3-L4 Please follow this VSCode friendly option

**Comment by TimothyMothra** [comment]
I'll need some help with the docker integration tests.

I added a new file for net7.
I can see that net7 was setup
```
Run actions/setup-dotnet@v[3](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:3).0.3
  with:
    dotnet-version: 7.0.x
/home/runner/work/_actions/actions/setup-dotnet/v3.0.3/externals/install-dotnet.sh --channel 7.0
...
dotnet-install: Attempting to download using aka.ms link https://dotnetcli.azureedge.net/dotnet/Sdk/7.0.100/dotnet-sdk-7.0.100-linux-x6[4](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:5).tar.gz
dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/7.0.100/dotnet-sdk-7.0.100-linux-x[6](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:7)4.tar.gz
dotnet-install: Installed version is [7](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:8).0.[10](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:11)0
```

but the run is failing because it's using net6
```
Step 8/12 : RUN dotnet publish "OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj" -c "${PUBLISH_CONFIGURATION}" -f "${PUBLISH_FRAMEWORK}" -o /drop -p:IntegrationBuild=true -p:TARGET_FRAMEWORK=${PUBLISH_FRAMEWORK}
 ---> Running in 044c292d5693
MSBuild version 17.3.2+561848881 for .NET
  Determining projects to restore...
Error: /usr/share/dotnet/sdk/6.0.403/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(144,5): error NETSDK1045: The current .NET SDK does not support targeting .NET 7.0.  Either target .NET 6.0 or lower, or use a version of the .NET SDK that supports .NET 7.0. [/repo/test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj]
The command '/bin/sh -c dotnet publish "OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj" -c "${PUBLISH_CONFIGURATION}" -f "${PUBLISH_FRAMEWORK}" -o /drop -p:IntegrationBuild=true -p:TARGET_FRAMEWORK=${PUBLISH_FRAMEWORK}' returned a non-zero code: 1
Service 'tests' failed to build : Build failed
```

**Comment by TimothyMothra** [comment]
> > I'll need some help with the docker integration tests.
> > I added a new file for net7. I can see that net7 was setup
> > ```
> > Run actions/setup-dotnet@v[3](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:3).0.3
> >   with:
> >     dotnet-version: 7.0.x
> > /home/runner/work/_actions/actions/setup-dotnet/v3.0.3/externals/install-dotnet.sh --channel 7.0
> > ...
> > dotnet-install: Attempting to download using aka.ms link https://dotnetcli.azureedge.net/dotnet/Sdk/7.0.100/dotnet-sdk-7.0.100-linux-x6[4](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:5).tar.gz
> > dotnet-install: Extracting zip from https://dotnetcli.azureedge.net/dotnet/Sdk/7.0.100/dotnet-sdk-7.0.100-linux-x[6](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:7)4.tar.gz
> > dotnet-install: Installed version is [7](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:8).0.[10](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/actions/runs/3424305346/jobs/5703828819#step:3:11)0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but the run is failing because it's using net6
> > ```
> > Step 8/12 : RUN dotnet publish "OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj" -c "${PUBLISH_CONFIGURATION}" -f "${PUBLISH_FRAMEWORK}" -o /drop -p:IntegrationBuild=true -p:TARGET_FRAMEWORK=${PUBLISH_FRAMEWORK}
> >  ---> Running in 044c292d5693
> > MSBuild version 17.3.2+561848881 for .NET
> >   Determining projects to restore...
> > Error: /usr/share/dotnet/sdk/6.0.403/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(144,5): error NETSDK1045: The current .NET SDK does not support targeting .NET 7.0.  Either target .NET 6.0 or lower, or use a version of the .NET SDK that supports .NET 7.0. [/repo/test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj]
> > The command '/bin/sh -c dotnet publish "OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.csproj" -c "${PUBLISH_CONFIGURATION}" -f "${PUBLISH_FRAMEWORK}" -o /drop -p:IntegrationBuild=true -p:TARGET_FRAMEWORK=${PUBLISH_FRAMEWORK}' returned a non-zero code: 1
> > Service 'tests' failed to build : Build failed
> > ```
> 
> Try setup for all versions in workflow - see if that works. https://github.com/open-telemetry/opentelemetry-dotnet/blob/d9e08386bb8038ae65a12719d5dcb83fe3f31898/.github/workflows/linux-ci.yml#L26-L32

I think that workflow is unrelated to the docker.
But that's good feedback, dotnet-setup steps should be combined :)

File: test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj [hunk 2]
```diff
@@ -1,7 +1,8 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (mic-max) [inline test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj:1]: whitespace change?

File: test/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests.csproj [hunk 3]
```diff
@@ -1,9 +1,14 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (mic-max) [inline test/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests.csproj:1]: nit whitespace change

**Review by mic-max — APPROVED** [review]
LGTM, extracting the bug fix as you said would be ideal 👍

File: test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj (hunk 2 above)
Comment (TimothyMothra) [reply test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj:1]: @mic-max Do you have any tips for how to view what changed here?
I didn't knowingly make any whitespace changes

File: test/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests.csproj (hunk 3 above)
Comment (TimothyMothra) [reply test/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests/OpenTelemetry.Instrumentation.EntityFrameworkCore.Tests.csproj:1]: Whis file, the encoding changed from UTF-8-BOM to UTF-8. 
Notepad++ shows this, as well as a hex editor.

I'll revert this change.

File: test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj (hunk 2 above)
Comment (TimothyMothra) [reply test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj:1]: I've looked at Notepad++, a hex editor, and WinMerge. I can't tell what changed here. :)

To get this resolved, i'll checkout the version from main and reapply my framework changes.

Comment (TimothyMothra) [reply test/OpenTelemetry.Extensions.AzureMonitor.Tests/OpenTelemetry.Extensions.AzureMonitor.Tests.csproj:1]: I made the changes and github still thinks something has changed on this file.

File: test/OpenTelemetry.Exporter.Stackdriver.Tests/OpenTelemetry.Exporter.Stackdriver.Tests.csproj [hunk 4]
```diff
@@ -1,7 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <Description>Unit test project for Stackdriver Exporter for OpenTelemetry</Description>
-    <TargetFrameworks>netcoreapp3.1;net6.0</TargetFrameworks>
+    <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
+    <TargetFrameworks>net7.0;net6.0;netcoreapp3.1</TargetFrameworks>
```
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.Stackdriver.Tests/OpenTelemetry.Exporter.Stackdriver.Tests.csproj:5]: would be good to remove netcoreapp3.1 as a separate PR givens its out of support soon.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**